compiler/transformation/transform.rs

```diff
@@ -61,3 +62,60 @@ fn prune_redundant_roleplayer_deduplication(conjunction: &mut Conjunction, block
         }
     });
 }
+
+pub fn optimize_away_statically_unsatisfiable_conjunctions(
+    conjunction: &mut Conjunction,
+    block_annotations: &TypeAnnotations,
+) {
+    optimize_away_statically_unsatisfiable_conjunctions_impl(conjunction, block_annotations);
+}
+
+fn optimize_away_statically_unsatisfiable_conjunctions_impl(
+    conjunction: &mut Conjunction,
+    block_annotations: &TypeAnnotations,
+) -> bool {
+    let mut must_optimise_away = false;
+    for nested in conjunction.nested_patterns_mut() {
+        match nested {
+            NestedPattern::Disjunction(disjunction) => {
+                let mut optimised_unsatisfiable_branch_ids = Vec::new();
+                for branch in disjunction.conjunctions_mut().iter_mut() {
+                    if optimize_away_statically_unsatisfiable_conjunctions_impl(branch, block_annotations) {
+                        optimised_unsatisfiable_branch_ids.push(branch.scope_id())
+                    }
+                }
+                disjunction.optimise_away_failing_branches(optimised_unsatisfiable_branch_ids);
+                must_optimise_away = must_optimise_away || disjunction.conjunctions().is_empty();
```

Comment:
We should break out here if `true`.

Reply:
We could.
I didn't so that all unsatisfiable sub-patterns would be replaced with an `OptimisedToUnsatisfiable`.